enhancements/network/dpu-ovnkube-operator.md

```diff
@@ -0,0 +1,186 @@
+---
+title: DPU OVNKube Operator
+authors:
+  - "@pliurh"
+reviewers:
+  - "@zshi-redhat"
+  - "@dcbw"
+approvers:
+  - 
+creation-date: 2021-09-07
+last-updated: 2021-09-07
+status: implementable
+---
+
+# DPU OVNKube Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+
+## Summary
+
+To facilitate the management of Nvidia BlueField-2 Smart-NIC, a two-cluster
+design is proposed. Under such design, a BlueFiled-2 card will be provisioned as
+a worker node of the ARM-based infra cluster, whereas the tenant cluster where
+the normal user applications run on, is composed of the X86 servers.
+
+The OVN-Kubernetes components are spread over the two clusters. On the tenant
+cluster side, the Cluster Network Operator is in charge of the management of the
+ovn-kube components. On the infra cluster side, We propose to create a new
+operator to be responsible for the life-cycle management of ovn-kube components.
+
+## Motivation
+
+Normally, the Cluster Network Operator watches the Network customer resource,
+then provision the ovn-kube components to the Openshift cluster.
+
+In the two-cluster design, some of the ovn-kube components have to run on the
+SmartNIC, but they are still part of the ovn-kube cluster running in the tenant
+cluster. To allow the ovn-kube components on different clusters to talk with
+each other, we need to ensure all the necessary configs and secrets are aligned
+with each other.
+
+### Goals
+
+- Design a new operator that
+  - manages the ovn-kube components that serve the tenant cluster.
+  - enable the swichedev mode on BlueField card.
+- Define the ovn-kube upgrade mechanism in such a deployment.
+
+### Non-Goals
+
+Support for network plugins other than OVN Kubernetes.
+This enhancement does not address the two-cluster installation.
+
+## Pre-requisites
+
+- This operator needs to access the apiserver of the tenant cluster, we need to
+make sure that the pods network in infra cluster can connect to the apiserver of
+the tenant cluster in some way.
+- One custom MachineConfigPool will be created for the BlueField nodes on which
+hardware offloading needs to be enabled.
+
+## Proposal
+
+We will create a namespaced CRD OVNKubeConfig.smartnic.openshift.io for the
+DPU OVNKube Operator. One custom resource of this CRD shall be created for
+the tenant cluster. The operator will take this custom resource as input to
```

Comment:
it's created "for the tenant cluster", but it's created _in_ the infra cluster, which again probably needs to be clarified.

How is the admin expected to create this resource? We should probably provide at least a script or something to grab the necessary information from the tenant cluster and create the resource in the infra cluster...

Reply:
I don't quite get the question. This operator only needs the kubeconfig of the tenant cluster, then it can access the apiserver to grab all the info. Do you mean we need a script to generate the kubeconfig file for the tenant cluster?

In my assumption, the whole process shall be like this:

1. After infra cluster installation, the admin labels the DPUs which he wants to enable hardware offload.
2. The admin creates a custom MCP that contains all the selected DPUs.
3. The admin configures the CNO to deprovision the ovn-kube components from the DPUs.
4. The admin installs the DPU operator and creates an OVNKubeConfig CR with the `poolName` but leaves `kubeConfigFile` blank. The operator then does the host network configuration.
5. Install the tenant cluster. Scale up the x86 workers with DPU.
6. The CNO of the tenant cluster puts those x86 hosts into smart-nic host mode.
7. The admin stores the kubeconfig file in the infra cluster then adds `kubeConfigFile` to the existing OVNKubeConfig CR. The operator installs the ovn-kube components on the selected DPUs.